### ISSUE TITLE
Excessive use of memory in some odd cases

### ISSUE THREAD

**Issue by glandium** [Tessil/robin-map#71]
I was trying to use heaptrack (a memory profiler) and it ended up using excessive amounts of memory. After investigation, this turned out to be from excessive use of memory by robin_set.
The original bug: https://bugs.kde.org/show_bug.cgi?id=477018
A reduced reproducer, with code extracted from heaptrack:
```
#include <iostream>
#include <cinttypes>
#include "robin_set.h"

struct IndexedAllocationInfo {
	uint64_t size;
	uint32_t index;

	bool operator==(const IndexedAllocationInfo& rhs) const {
            return rhs.index == index && rhs.size == size;
	}
};

// see: https://stackoverflow.com/a/2595226/35250
template <class T>
inline void hashCombine(std::size_t& seed, const T& v)
{
    std::hash<T> hasher;
    seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
}

namespace std {
template <>
struct hash<IndexedAllocationInfo>
{
    size_t operator()(const IndexedAllocationInfo& info) const
    {
        size_t seed = 0;
        hashCombine(seed, info.size);
        hashCombine(seed, info.index);
        return seed;
    }
};
}

int main() {
	tsl::robin_set<IndexedAllocationInfo> set;
	uint64_t size;
	uint32_t index;
	while (std::cin >> size >> index) {
		set.insert({size, index});
	}
	return 0;
}
```

The program reads a log of the data that was fed into the robin-set that used too much memory and replays it.
The log is [heaptrack.log.gz](https://github.com/Tessil/robin-map/files/13359845/heaptrack.log.gz)
Save the source above as `test.cc` and compile with `g++ -o test test.cc -I/path/to/include/tsl`.
Replay the log with `zcat heaptrack.log.gz | ./test`.
There are "only" slightly over 2M entries, and after about 470k, memory usage jumps above 40GB.
`std::unordered_set`, on the other hand, doesn't bat an eye.

It certainly doesn't help that the hash function doesn't produce very diverse values. Most are between 171800816639 and 180388095648, with 492913 being between 175100000000 and 175200000000, and another 1368894 between 175200000000 and 175300000000.

**Comment by milianw** [comment]
Using `boost::hash_combine`, the issue is gone - so I guess the hash combine I copied from stackoverflow is simply bad :)

**Comment by Tessil** [comment]
Yes, `tsl::robin_map` can behave quite badly when there is a high collision rate, trying to balance the load factor with the collision chain size. A rehash is forced when the collision chain reaches 8192, increasing the memory usage..
